wfe/wfe.go

```diff
@@ -1392,6 +1409,100 @@ func (wfe *WebFrontEndImpl) setCORSHeaders(response http.ResponseWriter, request
 	response.Header().Set("Access-Control-Max-Age", "86400")
 }
 
+type rolloverRequest struct {
+	OldKey jose.JsonWebKey
+	NewKey jose.JsonWebKey
+}
+
+// KeyRollover allows a user to change their signing key
+func (wfe *WebFrontEndImpl) KeyRollover(ctx context.Context, logEvent *requestEvent, response http.ResponseWriter, request *http.Request) {
+	body, oldKey, reg, prob := wfe.verifyPOST(ctx, logEvent, request, true, core.ResourceKeyChange)
+	addRequesterHeader(response, logEvent.Requester)
+	if prob != nil {
+		wfe.sendError(response, logEvent, prob, nil)
+		return
+	}
+
+	// Parse as JWS
+	newKey, parsedJWS, err := verifyJWSKey(string(body))
+	if err != nil {
+		logEvent.AddError(err.Error())
+		wfe.sendError(response, logEvent, probs.Malformed(err.Error()), err)
+		return
+	}
+	payload, err := parsedJWS.Verify(newKey)
+	if err != nil {
+		n := len(body)
+		if n > 100 {
+			n = 100
+		}
+		logEvent.AddError("verification of JWS with the JWK failed: %v; body: %s", err, body[:n])
+		wfe.sendError(response, logEvent, probs.Malformed("JWS verification error"), err)
+		return
+	}
+	var ror rolloverRequest
+	err = json.Unmarshal(payload, &ror)
+	if err != nil {
+		logEvent.AddError("unable to JSON parse resource from JWS payload: %s", err)
+		wfe.sendError(response, logEvent, probs.Malformed("Request payload did not parse as JSON"), nil)
+		return
+	}
+
+	// Check body contains the correct keys
+	pokThumb, err := ror.OldKey.Thumbprint(crypto.SHA256)
```

Comment:
There's no need to compare these by thumbprint, since we have full keys for both. Instead use [`MarshalPKIXPublicKey`](https://golang.org/pkg/crypto/x509/#MarshalPKIXPublicKey) and compare the bytes. We have a function `core.KeyDigestEquals` that does approximately this, but over digests. We should define an analogous `core.KeyEquals` that compares whole keys, and use that here.